File: tests/leap_compat/kotlin/CoreConsumer.kt
package consumer

import ai.liquid.leap.Conversation
import ai.liquid.leap.GenerationOptions
import ai.liquid.leap.ModelLoadingOptions
import ai.liquid.leap.ModelRunner
import ai.liquid.leap.message.ChatMessage
import ai.liquid.leap.message.MessageResponse
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.toList

fun stream(conversation: Conversation): Flow<MessageResponse> =
    conversation.generateResponse(
        userTextMessage = "hello",
        generationOptions =
            GenerationOptions.build {
                temperature = 0.2f
                rngSeed = 42L
                maxTokens = 16
            },
    )

suspend fun consume(conversation: Conversation): List<String> =
    stream(conversation).toList().map { response ->
        when (response) {
            is MessageResponse.Chunk -> response.text
            is MessageResponse.ReasoningChunk -> response.reasoning
            is MessageResponse.Error -> response.message
            is MessageResponse.Complete, is MessageResponse.AudioSample, is MessageResponse.FunctionCalls -> ""
        }
    }

fun loadingOptions(): ModelLoadingOptions = ModelLoadingOptions(contextSize = 2048, useMmap = false)

suspend fun history(
    runner: ModelRunner,
    messages: List<ChatMessage>,
): Int {
    val conversation: Conversation = runner.createConversationFromHistory(messages)
    conversation.removeLastMessage()
    val history: List<ChatMessage> = conversation.history
    return runner.getPromptTokensSize(history)
}

class RunnerConsumer(
    private val delegate: ModelRunner,
) : ModelRunner {
    override val modelId: String get() = delegate.modelId

    override fun createConversation(systemPrompt: String?): Conversation = delegate.createConversation(systemPrompt)

    override fun createConversationFromHistory(history: List<ChatMessage>): Conversation = delegate.createConversationFromHistory(history)

    override suspend fun unload() = delegate.unload()

    override suspend fun getPromptTokensSize(
        messages: List<ChatMessage>,
        addBosToken: Boolean,
    ): Int = delegate.getPromptTokensSize(messages, addBosToken)

    override suspend fun generateFromConversation(
        conversation: Conversation,
        callback: ModelRunner.GenerationCallback,
        generationOptions: GenerationOptions?,
    ): ModelRunner.GenerationHandler = delegate.generateFromConversation(conversation, callback, generationOptions)
}

class CallbackConsumer : ModelRunner.GenerationCallback {
    override fun onResponse(response: MessageResponse) {}

    override fun onError(error: Throwable) {}
}

class HandlerConsumer : ModelRunner.GenerationHandler {
    override fun stop() {}
}

File: tests/leap_compat/swift/CoreConsumer.swift
import Foundation
import LeapSDK

// Compile-only consumer: no model files or inference runtime are executed.
func stream(_ conversation: any Conversation) -> SkieSwiftFlow<any MessageResponse> {
  let options = GenerationOptions().with(temperature: 0.2).with(rngSeed: 42).with(maxTokens: 16)
  return conversation.generateResponse(userTextMessage: "hello", generationOptions: options)
}

func consume(_ conversation: any Conversation) async -> [String] {
  let responses: SkieSwiftFlow<any MessageResponse> = stream(conversation)
  var text: [String] = []
  for await response in responses {
    switch onEnum(of: response) {
    case .chunk(let chunk): text.append(chunk.text)
    case .reasoningChunk(let chunk): text.append(chunk.reasoning)
    case .error(let error): text.append(error.message)
    case .complete, .audioSample, .functionCalls: break
    }
  }
  return text
}

func history(_ runner: any ModelRunner, messages: [ChatMessage]) async throws -> KotlinInt {
  let conversation: any Conversation = runner.createConversationFromHistory(history: messages)
  conversation.removeLastMessage()
  let _: [ChatMessage] = conversation.history
  return try await runner.getPromptTokensSize(messages: messages, addBosToken: true)
}

final class CallbackConsumer: NSObject, ModelRunnerGenerationCallback {
  func onResponse(response: any MessageResponse) {}
  func onError(error: KotlinThrowable) {}
}

final class HandlerConsumer: NSObject, ModelRunnerGenerationHandler {
  func stop() {}
}
